File: propms/property_management_solution/page/tenant_dashboard/tenant_dashboard.py
import frappe
from datetime import datetime, timedelta

@frappe.whitelist()
def get_role_info():
    user = frappe.frappe.session.user
    user_doc = frappe.get_doc('User', user)
    return user_doc

@frappe.whitelist()
def balances():
    user = frappe.session.user
    user_doc = frappe.get_doc('User', user)
    
    today = datetime.today()
    first_day_next_month = (today.replace(day=1) + timedelta(days=32)).replace(day=1)
    last_day_next_month = (first_day_next_month + timedelta(days=32)).replace(day=1) - timedelta(days=1)

    first_day_next_month = first_day_next_month.date()
    last_day_next_month = last_day_next_month.date()

    leases = frappe.db.get_all('Lease', filters={"property_user": user_doc.full_name}, fields=["name"])

    seen_lease_items_next_month = set()
    seen_lease_items_all_upcoming = set()

    total_rate_next_month = 0
    total_rate_all_upcoming = 0

    currency_next_month = None
    currency_all_upcoming = None

    for lease in leases:
        schedules_next_month = frappe.db.get_all('Lease Invoice Schedule', 
            filters={
                'parent': lease.name, 
                'date_to_invoice': ['between', [first_day_next_month, last_day_next_month]]
            },
            fields=["name", "date_to_invoice", "rate", "lease_item_name", "currency"]
        )

        for schedule in schedules_next_month:
            lease_item_name = schedule.get('lease_item_name')
            if lease_item_name not in seen_lease_items_next_month:
                total_rate_next_month += schedule.get('rate', 0)
                currency_next_month = schedule.get('currency')
                seen_lease_items_next_month.add(lease_item_name)

        schedules_all_upcoming = frappe.db.get_all('Lease Invoice Schedule', 
            filters={
                'parent': lease.name, 
                'date_to_invoice': ['>=', first_day_next_month]
            },
            fields=["name", "date_to_invoice", "rate", "lease_item_name", "currency"]
        )

        for schedule in schedules_all_upcoming:
            lease_item_name = schedule.get('lease_item_name')
            if lease_item_name not in seen_lease_items_all_upcoming:
                total_rate_all_upcoming += schedule.get('rate', 0)
                currency_all_upcoming = schedule.get('currency')
                seen_lease_items_all_upcoming.add(lease_item_name)

    return {
        "upcoming_rent": total_rate_next_month,
        "all_upcoming_rent": total_rate_all_upcoming,
        "lease_item_names": list(seen_lease_items_all_upcoming),
        "currency": currency_next_month,
        "currency2": currency_next_month
    }

@frappe.whitelist()
def payments_history():
    user = frappe.session.user
    user_doc = frappe.get_doc('User', user)

    leases = frappe.db.get_all('Lease', filters={"property_user": user_doc.full_name}, fields=["name"])

    lease_schedules = []

    for lease in leases:
        schedules = frappe.db.get_all(
            'Lease Invoice Schedule', 
            filters={
                'parent': lease.name
            },
            fields=["parent", "date_to_invoice", "rate", "lease_item_name", "currency", "invoice_number"]
        )
        
        for schedule in schedules:
            if schedule.get('invoice_number'):
                lease_schedules.append(schedule)

    return lease_schedules

@frappe.whitelist()
def create_service(service, description, service_date):
    # Throwing service to test if it's received properly
    # frappe.throw(str(service))
    
    user = frappe.session.user  # Fetch the current logged-in user
    user_doc = frappe.get_doc('User', user)
    
    # Create a new 'Service Request' doc
    sq = frappe.new_doc("Service Request")
    sq.property_user = user_doc.full_name  # Assuming this is the user name
    sq.email = user_doc.email
    sq.phone_number = user_doc.mobile_no or user_doc.phone
    sq.address = user_doc.location or ""
    
    # Assign values from the function's arguments
    sq.service = service
    sq.service_date = service_date
    sq.description = description
    
    # Insert and save the new Service Request doc
    sq.insert(ignore_permissions=True)
    sq.save()

    return 'success'

File: propms/property_management_solution/page/tenant_dashboard/tenant_dashboard.js
frappe.pages['tenant-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Tenant Dashboard',
        single_column: true
    });

    // Render main HTML template
    page.main.html(frappe.render_template("tenant_dashboard", {}));
    page.add_action_item('Service Request', () => open_service_request_dialog());
    page.add_action_item('Online Payment', () => refresh());
    let $btn = page.set_secondary_action('Download', () => download_data())

    // Function to open the Service Request dialog (moved from set_primary_action)
    function open_service_request_dialog() {
        let d = new frappe.ui.Dialog({
            title: 'Service Request Details',
            fields: [
                {
                    label: 'Service',
                    fieldname: 'service',
                    fieldtype: 'Link',
                    options: 'Item',
                    get_query: function() {
                        return {
                            filters: {
                                'item_group': 'Services'
                            }
                        };
                    }
                },
                {
                    label: 'Description',
                    fieldname: 'description',
                    fieldtype: 'Small Text'
                },
                {
                    label: 'Service Date',
                    fieldname: 'service_date',
                    fieldtype: 'Date'
                }
            ],
            primary_action_label: 'Submit',
            primary_action(values) {
                console.log(values);

                // Validate form fields (optional but recommended)
                if (!values.service || !values.service_date || !values.description) {
                    frappe.msgprint(__('Please fill out all fields before submitting.'));
                    return;
                }

                frappe.call({
                    method: 'propms.property_management_solution.page.tenant_dashboard.tenant_dashboard.create_service',
                    args: {
                        service: values.service,
                        description: values.description,
                        service_date: values.service_date
                    },
                    callback: (r) => {
                        if (r.message === 'success') {
                            frappe.show_alert({
                                message: ('Hi, Service request created successfully.'),
                                indicator: 'green'
                            }, 5);
                        } else {
                            frappe.throw(__('There was an issue creating the service request.'));
                        }
                    },
                    error: (r) => {
                        frappe.throw(__('An error occurred while creating the service request.'));
                    }
                });
                d.hide();
            }
        });

        d.show();
    }

    function download_data() {
        frappe.call({
            method: 'propms.property_management_solution.page.tenant_dashboard.tenant_dashboard.payments_history',
            callback: (response) => {download_data
                let payments_history = response.message;
                let csvContent = "data:text/csv;charset=utf-8,";
                csvContent += "Invoice Number, Date to Invoice, Lease Item Name, Rate, Currency\n"; // CSV Headers
                
                // Convert the payments history to CSV format
                payments_history.forEach((payment) => {
                    let row = [
                        payment.invoice_number || '-',
                        payment.date_to_invoice || '-',
                        payment.lease_item_name || '-',
                        payment.rate ? payment.rate.toLocaleString() : '-',
                        payment.currency || '-'
                    ];
                    csvContent += row.join(",") + "\n";
                });

                // Create a link element for download
                var encodedUri = encodeURI(csvContent);
                var link = document.createElement("a");
                link.setAttribute("href", encodedUri);
                link.setAttribute("download", "payments_history.csv");
                document.body.appendChild(link); // Required for Firefox

                // Trigger the download
                link.click();

                // Remove the link element after triggering download
                document.body.removeChild(link);
            }
        });
    }


    // Fetch user data
    frappe.call({
        method: 'propms.property_management_solution.page.tenant_dashboard.tenant_dashboard.get_role_info',
        callback: (r) => {
            let user = r.message;
            $("#user-image").attr("src", user.user_image);
            $("#user-name").text(user.full_name);
            $("#user-email").text(user.email || "-");
            $("#user-location").text(user.location || "-");
            $("#user-phone").text(user.phone || "-");
            $("#user-username").text(user.username || "-");
            $("#user-gender").text(user.gender || "-");
        },
        error: (r) => {
            console.error("Error fetching user data:", r);
        }
    });

    // Fetch balances
    frappe.call({
        method: 'propms.property_management_solution.page.tenant_dashboard.tenant_dashboard.balances',
        callback: (response) => {
            let lease = response.message;

            let upcomingRent = lease.upcoming_rent ? lease.upcoming_rent.toLocaleString() : "-";
            let allUpcomingRent = lease.all_upcoming_rent ? lease.all_upcoming_rent.toLocaleString() : "-";

            $("#upcoming_rent").text(upcomingRent);
            $("#all_upcoming_rent").text(allUpcomingRent);
            $("#lease_item_names").text(lease.lease_item_names || "-");
            $("#currency").text(lease.currency || "-");
            $("#currency2").text(lease.currency2 || "-");

            // console.log(upcomingRent, allUpcomingRent);
        }
    });

    // Fetch payment history
    frappe.call({
        method: 'propms.property_management_solution.page.tenant_dashboard.tenant_dashboard.payments_history',
        callback: (response) => {
            let payments_history = response.message;
            // console.log(payments_history);

            let tableBody = document.getElementById('payments-history-body');
            tableBody.innerHTML = "";

            payments_history.forEach((payment) => {
                let row = document.createElement('tr');
                row.innerHTML = `
                    <td>${payment.invoice_number || '-'}</td>
                    <td>${payment.date_to_invoice || '-'}</td>
                    <td>${payment.lease_item_name || '-'}</td>
                    <td>${payment.rate ? payment.rate.toLocaleString() : '-'}</td>
                    <td>${payment.currency || '-'}</td>
                `;
                tableBody.appendChild(row);
            });
        }
    });
};
